File: distrebuteCoins/distrebuteCoin.cpp
#include <iostream>
#include <cstdlib>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int distributeCoins(TreeNode* root) {
        int moves = 0;
        dfs(root, moves);
        return moves;
    }

private:
    int dfs(TreeNode* node, int& moves) {
        if (!node) return 0;

        int left_excess = dfs(node->left, moves);
        int right_excess = dfs(node->right, moves);

        moves += abs(left_excess) + abs(right_excess);

        return node->val + left_excess + right_excess - 1;
    }
};

int main() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(0);
    root->right = new TreeNode(0);

    Solution sol;
    cout << "Minimum moves: " << sol.distributeCoins(root) << endl;

    delete root->left;
    delete root->right;
    delete root;

    return 0;
}

File: distrebuteCoins/distrebuteCoin.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class Solution:
    def distributeCoins(self, root: TreeNode) -> int:
        moves = [0]
        self.dfs(root, moves)
        return moves[0]

    def dfs(self, node: TreeNode, moves: list) -> int:
        if not node:
            return 0

        left_excess = self.dfs(node.left, moves)
        right_excess = self.dfs(node.right, moves)

        moves[0] += abs(left_excess) + abs(right_excess)

        return node.val + left_excess + right_excess - 1

# Tạo cây nhị phân từ ví dụ 1
root = TreeNode(3)
root.left = TreeNode(0)
root.right = TreeNode(0)

sol = Solution()
print("Minimum moves:", sol.distributeCoins(root))
